### ISSUE TITLE
Double operation cancel completion call

### ISSUE THREAD

**Issue by anton-plebanovich** [SDWebImage/SDWebImage#3374]
### Issue Info

 Info                    | Value                               |
-------------------------|-------------------------------------|
 Platform Name           | ios
 Platform Version        | 15.5
 SDWebImage Version      | 5.13.0
 Integration Method      | cocoapods
 Xcode Version           | Xcode 13.4.1
 Repro rate              | sometimes
 Repro with our demo prj | -
 Demo project link       | -

### Issue Description and Steps

After we updated to `5.13.0` from `5.12.5` the completion started to fire twice when new image is set while there is an ongoing set operation. 

The first stack trace
<img width="969" alt="Screenshot 2022-07-19 at 15 28 40" src="https://user-images.githubusercontent.com/10807812/179751242-7aa70e1e-ec74-4984-9e62-6d314df151c5.png">

The second stack trace
<img width="990" alt="Screenshot 2022-07-19 at 15 33 39" src="https://user-images.githubusercontent.com/10807812/179751506-5bbc3c8c-2293-4bff-8c88-4b2594629276.png">

**Comment by dreampiggy** [comment]
From the desired design from the beginning, when you cancel the image loading, it should not trigger the completion block callback for UIImageView loading block.

This seems introduced during v5.13.1

**Comment by anton-plebanovich** [comment]
Thanks for your response and the hard work

**Comment by dreampiggy** [comment]
@anton-plebanovich
Hey. Isn't that you who fired https://github.com/SDWebImage/SDWebImage/issues/3141 ?

At that time SDWebImage v4 any `cancel` will never callback. And we change it to always callback in v5 😂

**Comment by dreampiggy** [comment]
The only differences between v5.12 and v5.13 is whether the `cancel` callback is sync or async.

Code:

```swift
imageView.sd_setImage(with: url1, placeholderImage: placeholder)
imageView.sd_setImage(with: url2, placeholderImage: placeholder)
```

### 5.12 Behavior:

1. imageView.image = placeholder (sync)
2. imageView.image = placeholder (sync)
3. imageView.image = nil (aync)
4. imageView.image = url2image (async)

### 5.13 Behavior:

1. imageView.image = placeholder (sync)
2. imageView.image = nil (sync)
3. imageView.image = placeholder (sync)
4. imageView.image = url2image (async)

**Comment by anton-plebanovich** [comment]
> @anton-plebanovich Hey. Isn't that you who fired #3141 ?
> 
> At that time SDWebImage v4 any `cancel` will never callback. And we change it to always callback in v5 😂

Well yes, we expect a callback for each image set operation. There is a logic that relies on a callback even if that's a cancel but it should be fired only once so when we receive a callback twice it causes some issues. Though, it's better to receive it twice than zero times. We are filtering excessive calls as a workaround for now.

**Comment by dreampiggy** [comment]
> The only differences between v5.12 and v5.13 is whether the `cancel` callback is sync or async.
> 
> Code:
> 
> ```swift
> // assume url1 and url2 is not cached
> imageView.sd_setImage(with: url1, placeholderImage: placeholder)
> imageView.sd_setImage(with: url2, placeholderImage: placeholder)
> ```
> 
> ### 4.0 Behavior:
> UI(setImage):
> 
> 1. imageView.image = placeholder (sync)
> 2. imageView.image = placeholder (sync)
> 3. imageView.image = url2image (async)
> 
> Callback:
> 
> 1. callback - url2 (async)
> 
> ### 5.12 Behavior:
> UI(setImage):
> 
> 1. imageView.image = placeholder (sync)
> 2. imageView.image = placeholder (sync)
> 3. imageView.image = nil (async - url1 cancelled)
> 4. imageView.image = url2image (async)
> 
> Callback:
> 
> 1. callback - url1 (async)
> 2. callback - url2 (async)
> 
> This may cause imagView rendering blank and then replace with url2image, like visual flash
> 
> ### 5.13 Behavior:
> UI(setImage):
> 
> 1. imageView.image = placeholder (sync)
> 2. imageView.image = nil (sync - url1 cancelled)
> 3. imageView.image = placeholder (sync)
> 4. imageView.image = url2image (async)
> 
> Callback:
> 
> 1. callback - url1 (sync)
> 2. callback - url2 (async)
> 
> ### Future ? 5.13.2 Behavior:
> UI(setImage):
> 
> 1. imageView.image = placeholder (sync)
> 2. imageView.image = placeholder (sync)
> 3. imageView.image = url2image (async)
> 
> Callback: 2. callback - url2 (async)

Does this match your desired behavior ?

**Comment by anton-plebanovich** [comment]
> Does this match your desired behavior ?
I think what we observe at the moment is:

5.13 Behavior:
...
Callback:
callback - url1 (sync - cancel)
callback - url1 (async - cancel)
callback - url2 (async - set)

The issue is that we receive `callback - url1` for a `cancel` twice

**Comment by dreampiggy** [comment]
I know what cause this, 5.13 only fix the cache query cancel, but now download cancel.

When cancel on a operation which already submitted to `SDWebImageDownloader`, it not sync but async.

This cause the second unwanted `callback - url1 (async - cancel)`

**Comment by dreampiggy** [comment]
@anton-plebanovich

Hi.
Actually, any demo code or minimum reproduce demo can help me to test ?

I check your backtrace again, it actually does not match what I think the issue is. These two stacktraces are not on single thread state, correct ?

Your first stacktrace is designed behavior in future, because you cancel an exist operation, which previous image is querying memory cache **synchronously**, it finally will callback your previous image loading completion block in synchronously.

Your second stacktrace is another cancellation, which previous image is querying disk cache **asynchronously** and seems finished ? It finally callback your previous image loading completion block in asynchronously.

**Comment by anton-plebanovich** [comment]
Let me some time to reproduce the issue in the playground so I can share the code

**Comment by dreampiggy** [comment]
I know, maybe this because of this line. GCD dispatch need time

<img width="950" alt="截屏2022-07-23 21 00 35" src="https://user-images.githubusercontent.com/6919743/180606088-a4c16dbc-dc8f-4e83-b9fc-0890111ceb98.png">

**Comment by anton-plebanovich** [comment]
@dreampiggy Here is a repro you may play with https://github.com/anton-plebanovich/SingleView/tree/bug/sd-web-image-double-cancel
```
Called error completion for 'Octocat.png': Error Domain=SDWebImageErrorDomain Code=2002 "Operation cancelled by user during querying the cache" UserInfo={NSLocalizedDescription=Operation cancelled by user during querying the cache}
Called error completion for 'Octocat.png': Error Domain=SDWebImageErrorDomain Code=2002 "Operation cancelled by user during querying the cache" UserInfo={NSLocalizedDescription=Operation cancelled by user during querying the cache}
Called success completion for 'GitHub-Mark.png'
```

**Comment by dreampiggy** [comment]
Released v5.13.2 for this.

**Comment by huoda1237** [comment]
Error Domain=SDWebImageErrorDomain Code=2002 "Operation cancelled by user during sending the request" UserInfo={NSLocalizedDescription=Operation cancelled by user during sending the request


in  v5.15.5

**Comment by dreampiggy** [comment]
@huoda1237 Create a new issue here and maybe provide a demo for me ?

https://github.com/SDWebImage/SDWebImage/issues
